### PR TITLE
CASSANDRA-19688 SAI support for BETWEEN operator

### PR DESCRIPTION
**Pull request by arkn98**
Patch by Arun Ganesh; reviewed by TBD for CASSANDRA-19688

### CODE REVIEW

File: src/java/org/apache/cassandra/index/sai/plan/Expression.java [hunk 1]
```diff
@@ -208,6 +211,27 @@ public Expression add(Operator op, ByteBuffer value)
...
+                if (indexTermType.isReversed())
+                {
+                    upper = new Bound(buffers.get(0), indexTermType, true);
+                    lower = new Bound(buffers.get(1), indexTermType, true);
+                }
+                else
+                {
+                    lower = new Bound(buffers.get(0), indexTermType, true);
+                    upper = new Bound(buffers.get(1), indexTermType, true);
+                }
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/index/sai/plan/Expression.java]: An alternate way to express this, if you think it reads better (not binding by any means) is...

```
int lowerIndex = indexTermType.isReversed() ? 1 : 0;
int upperIndex = indexTermType.isReversed() ? 0 : 1;
lower = new Bound(buffers.get(lowerIndex), indexTermType, true);
upper = new Bound(buffers.get(upperIndex), indexTermType, true);
```

**Review by maedhroz — APPROVED** [review]
LGTM, let's just make sure to cover the reversed type case in a test somewhere ;)

File: test/unit/org/apache/cassandra/index/sai/cql/types/IndexingTypeSupport.java [hunk 2]
```diff
@@ -74,7 +74,7 @@ public void createTable()
     {
         dataset.init();
 
-        createTable(String.format("CREATE TABLE %%s (pk int, ck int, value %s, PRIMARY KEY(pk, ck))", dataset));
+        createTable(String.format("CREATE TABLE %%s (pk int, ck int, value %s, PRIMARY KEY(pk, ck)) WITH CLUSTERING ORDER BY (ck desc)", dataset));
```
Comment (arkn98) [inline test/unit/org/apache/cassandra/index/sai/cql/types/IndexingTypeSupport.java]: I'm not sure if this is the right place to test the `indexTermType.isReversed()` logic.

Comment (arkn98) [reply test/unit/org/apache/cassandra/index/sai/cql/types/IndexingTypeSupport.java]: And, I've yet to modify `RandomIntersectionTester`.

Comment (maedhroz) [reply test/unit/org/apache/cassandra/index/sai/cql/types/IndexingTypeSupport.java]: Resolved both of these on Slack

File: src/java/org/apache/cassandra/db/filter/RowFilter.java [hunk 3]
```diff
@@ -383,6 +383,30 @@ public RowFilter without(ColumnMetadata column, Operator op, ByteBuffer value)
         return withNewExpressions(newExpressions);
     }
 
+    /**
+     * @return a copy of this filter without any primary key expressions, or simply itself if non exist already
+     */
+    public RowFilter withoutKeyExpressions()
```
Comment (jasonstack) [inline src/java/org/apache/cassandra/db/filter/RowFilter.java]: I don't think this is the right approach to fix BETWEEN#isSatified issue.. This does not fix root cause

**Review by jasonstack — CHANGES_REQUESTED** [review]
in my understanding, the root cause of BETWEEN failure in SAI is `BETWEEN#isSatifiedBy` doesn't use the correct lower bound and upper bound.

How about fixing the `BETWEEN#isSatifiedBy` directly without skipping key expression? we need some unit tests around it
```
        @Override
        public boolean isSatisfiedBy(AbstractType<?> type, ByteBuffer leftOperand, ByteBuffer rightOperand)
        {
            List<ByteBuffer> buffers = ListType.getInstance(type, false).unpack(rightOperand);
            buffers.sort(type);
            int lowerBoundIdx = type.isReversed() ? 1 : 0;
            int upperBoundIdx = type.isReversed() ? 0 : 1;
            return type.compareForCQL(leftOperand, buffers.get(lowerBoundIdx)) >= 0 && type.compareForCQL(leftOperand, buffers.get(upperBoundIdx)) <= 0;
        }
```

File: src/java/org/apache/cassandra/db/filter/RowFilter.java (hunk 3 above)
Comment (jasonstack) [reply src/java/org/apache/cassandra/db/filter/RowFilter.java]: Talked to Caleb, it makes sense to skip RFP if only primary key columns are filtered

Comment (maedhroz) [reply src/java/org/apache/cassandra/db/filter/RowFilter.java]: BETWEEN itself now fixed via [CASSANDRA-19878](https://github.com/blerer/cassandra/tree/CASSANDRA-19878)

File: src/java/org/apache/cassandra/index/sai/plan/Expression.java [hunk 4]
```diff
@@ -208,6 +211,20 @@ public Expression add(Operator op, ByteBuffer value)
...
+
+            case BETWEEN:
+                ListType<?> type = ListType.getInstance(indexTermType.columnMetadata().type, false);
+                List<? extends ByteBuffer> buffers = type.unpack(value);
+
+                operator = IndexOperator.RANGE;
+                this.upperInclusive = true;
+                this.lowerInclusive = true;
+
+                lower = new Bound(buffers.get(0), indexTermType, true);
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/index/sai/plan/Expression.java:223]: I'll add a comment here on commit. Something like...

```
// SimpleRestriction#addToRowFilter() ensures correct buffer ordering even with ReversedType
```

**Review by jasonstack — APPROVED** [review]
LGTM

**Comment by maedhroz** [comment]
Committed as https://github.com/apache/cassandra/commit/bddaa4409fc4cb83394c2963a1028bcfb9dce733

**Pull request closed without merging**